docs/sql/recursive.md

```diff
@@ -0,0 +1,164 @@
+
+Finally, given a view forward declaration, views can be defined using
+the standard SQL syntax using arbirary query expressions.  The forward
+declaration allows the views to write queries that depend on the
+views themselves:
+
+```sql
+CREATE TABLE EDGES(x int, y int);
+
+-- a local view that depends  on CLOSURE and EDGES
+CREATE LOCAL VIEW STEP AS
+SELECT E.x, CLOSURE.y FROM
+E JOIN CLOSURE ON e.y = CLOSURE.x
+
+-- actual definition of view CLOSURE, which depends on STEP and E
+CREATE MATERIALIZED VIEW CLOSURE AS
+(SELECT * FROM EDGES) UNION (SELECT * FROM STEP);
+```
+
+Notice that view `STEP` does not need a forward declaration, for the
+same reason that function `f` in the C example does not need one.
+However, if view `STEP` is not a `LOCAL` view, then it does need a
+forward declaration -- all views that are user-visible that are
+mutually recursive need forward declarations.
+
+The type inferred by the compiler for the view based on the query must
+match exactly the type of the declared view, including the nullability
+of the columns, and the column names.
+
+### Semantics
+
+Currently the compiler automatically makes all mutually-recursive
+views `DISTINCT` (they cannot contain duplicates).
+
+The semantics of mutually-recursive views matches the semantics from
+other programming languages: for any contents of the input tables, the
+views initially start empty, and then the queries for the views are
+evaluated until the views stop changing.  Thus a set of mutually
+recursive views computes the least fixpoint of the queries.
+
+(Note that this is *not* the semantics that SQL gives to recursive
+queries, which is considered broken.)
+
+Moreover, in incremental mode, when tables are updated using
+insertions or deletions, the views are automatically updated
+incrementally to reflect the latest contents of the tables.
+
+In the above SQL example the `EDGES` table represents the edges of a
+graph where nodes are integers.  The `CLOSURE` view represents the
+edges of a graph over the same nodes, such that `CLOSURE` is the
+transitive closure of the `EDGES` graph.
+
+We can verify this:
+
+```sql
+INSERT INTO EDGES VALUES(0, 1);
+SELECT * FROM CLOSURE;
+ x | y
+-------
+ 0 | 1
+(1 row)
+INSERT INTO EDGES VALUES(1, 2);
+SELECT * FROM CLOSURE;
+ x | y
+-------
+ 0 | 1
+ 1 | 2
+ 0 | 2
+(3 rows)
+DELETE * FROM EDGES WHERE x = 0;
+SELECT * FROM CLOSURE;
+ x | y
+-------
+ 1 | 2
+(1 row)
+```
+
+### Restrictions
+
+Feldera supports a much wider range of operators in recursive queries
+than other SQL dialects.  Currently the following operators *can* be
+used in recursive queries: `SELECT`, `WHERE`, `GROUP BY`, all kinds of
+`JOIN`s, `HAVING`, aggregations, `DISTINCT`, `UNNEST`, `VALUES`,
+`ORDER BY`, `PIVOT`, `UNION`, `INTERSECT`, `EXCEPT`, table functions
+such as `HOP` and `TUMBLE`.
+
+Unsupported constructs include window functions (using `OVER`).
+
+Programs can mix [streaming computations](streaming) and recursive
```

Comment:
we don't really have streaming computations. We have some annotations used for GC.